ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -251,3 +260,65 @@ longestCrowdedRun (LeaderSchedule m) = fst
         | n == 0    = (x, y)
         | n == 1    = (x, noRun)
         | otherwise = let y' = incCrowdedRun slot y in (max x y', y')
+
+{-------------------------------------------------------------------------------
+  Node Connection Schedule - when each node joins the test network
+-------------------------------------------------------------------------------}
+
+-- | A list of durations as number of slots.
+-- Each duration is the amount of time after a node joins the network before
+-- either the next node does or the test ends.
+--
+-- ASSUMPTIONS:
+-- * Nodes join in the order of their Id.
+-- * Node 0 joins first, at slot 0, so the first duration in the list occurs after that.
+-- * Nodes other than Node 0 join after some possibly empty duration
+--   since the previous node joined.
+-- * The last duration in the list is the run duration after the last node joins.
+-- * The number of durations is therefore the number of nodes.
+newtype ConnectionPlan = ConnectionPlan [Int]
+      -- ^
+      --
+      -- INVARIANT: all non-negative.
+      --
+      -- For example, for a test with 4 nodes and 10 slots, the durations could be @[0,6,1,2]@
+      -- if the first two join simultaneously at slot 0,
+      -- the third joins at slot 6,
+      -- the fourth joins at slot 7,
+      -- the full net runs for 2 more slots,
+      -- and the test ends at slot 9 (i.e. the 10th slot).
+    deriving (Show, Eq)
+
+connectionPlanNumSlots :: ConnectionPlan -> NumSlots
+connectionPlanNumSlots (ConnectionPlan durs) = NumSlots (sum durs)
+
+connectionPlanNumCoreNodes :: ConnectionPlan -> NumCoreNodes
+connectionPlanNumCoreNodes (ConnectionPlan durs) = NumCoreNodes (length durs)
+
+-- | When each node joins (Core nodes Ids @0 .. n - 1@),
+-- and the total number of slots in the simulation.
+interpretConnectionPlan :: ConnectionPlan -> ([SlotNo],NumSlots)
+interpretConnectionPlan (ConnectionPlan durs) = (joinSlots,numSlots)
+  where
+    (durSum,joinSlots) = mapAccumL
+        (\acc dur -> (acc + dur,SlotNo (toEnum acc)))
+        0 durs
+    numSlots = NumSlots durSum
+
+-- | The trivial \"plan\" where all nodes join immediately in slot 0.
+immediateJoinConnectionPlan :: NumCoreNodes -> NumSlots -> ConnectionPlan
+immediateJoinConnectionPlan (NumCoreNodes nnode) (NumSlots nslot) =
+  ConnectionPlan $ replicate (nnode - 1) 0 ++ [nslot]
+
+-- | To avoid redundant shrinking of 'ConnectionPlan',
+-- we carry some additional information used only to guide shrinking.
+data ShrunkConnectionPlan = ShrunkConnectionPlan
```

Comment:
Skipping this on your request.

It might indeed be useful to store additional information to improve shrinking, e.g., something like a "stage": we are shrinking `NumCoreNodes`, `NumSlots`, or individual `SlotNo`s, etc.